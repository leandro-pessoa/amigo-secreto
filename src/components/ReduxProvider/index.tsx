// componentes
import { Provider } from 'react-redux'

// store
import { setupStore } from '@/app/store'

// tipagens
import { ReactElement } from 'react'

// tipagem dos props
interface ReduxProviderProps {
    children: ReactElement | ReactElement[] | string
}

const ReduxProvider = ({ children }: ReduxProviderProps) => {
    return <Provider store={setupStore()}>{children}</Provider>
}

export default ReduxProvider
