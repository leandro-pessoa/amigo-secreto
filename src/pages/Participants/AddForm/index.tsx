// funções e tipagens
import { useAppDispatch, useAppSelector } from '@/app/hooks'
import React, { useState } from 'react'
import { error } from '@/utils/feedbacks'
import { v4 as uuidv4 } from 'uuid'
import { useReloadDrawn } from '@/app/hooks/useReloadDrawn'

// tipagens externas
import { IParticipant } from '@/interfaces/IParticipant'

// actions
import { addParticipant } from '@/app/reducers/participantReducer'

// componentes
import { StyledForm } from './styles'

// states globai
import { selectParticipants } from '@/app/reducers/participantReducer'

const AddForm = () => {
    // states globais
    const dispatch = useAppDispatch()
    const participants = useAppSelector(selectParticipants)

    // states
    const [name, setName] = useState<string>('')

    // constantes
    const reloadDrawn = useReloadDrawn()

    // adiciona o participante à lista caso não seja repetido
    // e o nome tenha 3 ou mais letras
    const addHandle = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault()

        const participant: IParticipant = {
            id: uuidv4(),
            name: name.trim(),
        }
        const participantsNames = participants.map((p) => p.name)

        if (name.trim().length < 3) {
            error('Um nome deve conter 3 ou mais letras.')
            return
        }
        if (participantsNames.includes(name)) {
            error(`O nome ${name} já foi adicionado.`)
            return
        }
        dispatch(addParticipant(participant))
        setName('')
        reloadDrawn()
    }

    return (
        <StyledForm onSubmit={(e) => addHandle(e)}>
            <input
                type='text'
                placeholder='Inserir participante'
                value={name}
                onChange={(e) => setName(e.target.value)}
                maxLength={15}
            />
            <button type='submit' disabled={name ? false : true}>
                Adicionar
            </button>
        </StyledForm>
    )
}

export default AddForm
