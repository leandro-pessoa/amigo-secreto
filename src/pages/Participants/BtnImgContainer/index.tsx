// funções
import { useNavigate } from 'react-router-dom'

// componentes
import Button from '@/components/Button'
import { StyledSection } from './styles'

// imagens
import bags from '/imagens/sacolas.png'

// ícones
import { FaRegCirclePlay } from 'react-icons/fa6'

const BtnImgContainer = () => {
    // navegação
    const navigate = useNavigate()

    return (
        <StyledSection>
            <Button onClick={() => navigate('/sortear')}>
                <FaRegCirclePlay size={26} />
                Iniciar brincadeira!
            </Button>
            <img src={bags} alt='Imagem de duas sacolas de compras' />
        </StyledSection>
    )
}

export default BtnImgContainer
