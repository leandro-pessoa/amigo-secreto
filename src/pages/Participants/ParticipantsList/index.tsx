// funções
import { useAppDispatch, useAppSelector } from '@/app/hooks'
import { participants } from '@/utils/participants'
import { sort } from '@/utils/sort'

// componentes
import { StyledDiv } from './styles'
import ParticipantItem from './ParticipantItem'
import Filters from './Filters'
import Container from '@/components/Container'
import SubTitle from '@/components/SubTitle'
import ReloadButton from '@/components/ReloadButton'

// actions
import { changeModalType, toggleModalDisplay } from '@/app/reducers/appReducer'
import { setParticipants, showAll } from '@/app/reducers/participantReducer'

// states globais
import { selectParticipants } from '@/app/reducers/participantReducer'
import { selectSelectedOption } from '@/app/reducers/appReducer'

const PaticipantsList = () => {
    // states globais
    const participantsState = useAppSelector(selectParticipants)
    const selectedOption = useAppSelector(selectSelectedOption)
    const dispatch = useAppDispatch()

    const removeAllModalHandle = () => {
        dispatch(changeModalType('removeAll'))
        dispatch(toggleModalDisplay(true))
    }

    const showAllHandle = () => {
        dispatch(showAll())
        dispatch(setParticipants(sort(participants(), selectedOption)))
    }

    return (
        <StyledDiv>
            {participantsState.length >= 1 ? (
                <>
                    <SubTitle>Participantes</SubTitle>
                    <Filters />
                    <Container>
                        <div className='list__btn-container'>
                            {participants().length ===
                            participantsState.length ? (
                                <></>
                            ) : (
                                <ReloadButton
                                    onClick={() => showAllHandle()}
                                    size={26}
                                    className='btn-container__show-all'
                                />
                            )}
                            <button
                                onClick={() => removeAllModalHandle()}
                                className='btn-container__remove-all'
                            >
                                Remover todos
                            </button>
                        </div>
                        <ul className='participants__list'>
                            {participantsState.map((participant) => (
                                <ParticipantItem
                                    key={participant.id}
                                    {...participant}
                                />
                            ))}
                        </ul>
                    </Container>
                </>
            ) : (
                <SubTitle>Não há participantes na lista</SubTitle>
            )}
        </StyledDiv>
    )
}

export default PaticipantsList
