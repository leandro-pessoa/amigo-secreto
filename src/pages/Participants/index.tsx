// componentes
import AddForm from '@/pages/Participants/AddForm'
import Title from '@/components/Title'
import PaticipantsList from '@/pages/Participants/ParticipantsList'
import BtnImgContainer from './BtnImgContainer'

const Participants = () => {
    return (
        <>
            <Title>Vamos come√ßar!</Title>
            <AddForm />
            <PaticipantsList />
            <BtnImgContainer />
        </>
    )
}

export default Participants
