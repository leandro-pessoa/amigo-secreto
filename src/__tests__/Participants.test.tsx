// funções
import { render } from '@testing-library/react'

// componentes
import ReduxProvider from '@/components/ReduxProvider'
import Participants from '../pages/Participants'

const mockNavigate = vi.fn()

vi.mock('react-router-dom', () => {
    return {
        useNavigate: () => mockNavigate,
    }
})

describe('a pagina de participantes', () => {
    test('deve ser renderizada corretamente', () => {
        const { container } = render(
            <ReduxProvider>
                <Participants />
            </ReduxProvider>,
        )

        expect(container).toMatchSnapshot()
    })
})
