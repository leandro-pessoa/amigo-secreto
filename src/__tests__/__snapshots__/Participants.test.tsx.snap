// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`a pagina de participantes > deve ser renderizada corretamente 1`] = `
<div>
  <h1
    class="sc-guDLey iEsuFB"
  >
    Vamos começar!
  </h1>
  <form
    class="sc-beySPh bjxaAz"
  >
    <input
      maxlength="15"
      placeholder="Inserir participante"
      type="text"
      value=""
    />
    <button
      disabled=""
      type="submit"
    >
      Adicionar
    </button>
  </form>
  <div
    class="sc-dmyCSP kNXCQg"
  >
    <h2
      class="sc-qZrbh bOgRFK"
    >
      Não há participantes na lista
    </h2>
  </div>
  <section
    class="sc-irLvIq eqXVxk"
  >
    <button
      class="sc-kFCroH grXOZX"
    >
      <svg
        fill="currentColor"
        height="26"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="26"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M464 256A208 208 0 1 0 48 256a208 208 0 1 0 416 0zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zM188.3 147.1c7.6-4.2 16.8-4.1 24.3 .5l144 88c7.1 4.4 11.5 12.1 11.5 20.5s-4.4 16.1-11.5 20.5l-144 88c-7.4 4.5-16.7 4.7-24.3 .5s-12.3-12.2-12.3-20.9V168c0-8.7 4.7-16.7 12.3-20.9z"
        />
      </svg>
      Iniciar brincadeira!
    </button>
    <img
      alt="Imagem de duas sacolas de compras"
      src="/imagens/sacolas.png"
    />
  </section>
</div>
`;
